<a href="https://colab.research.google.com/github/Pranav-Tech46/Machine-Learning-Practice/blob/main/Assessment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [2]:
df = pd.read_csv('/content/tough_dataset - Copy2.csv')
df.head()


id  Target A  ...  next_pymnt_d  last_credit_pull_d
0  68407277  42405.00  ...           NaN              19-Mar
1  68355089  60740.00  ...           NaN              19-Mar
2  68341763  51907.68  ...           NaN              19-Mar
3  66310712  51907.68  ...        19-Apr              19-Mar
4  68476807     35.00  ...           NaN              18-Mar

[5 rows x 69 columns]

In [3]:
df.isnull().sum().head(10)


id                             0
Target A                       0
Target B                       0
LOAN_AMT                       0
Business Title                 0
Civil Service Title            0
Division/Work Unit             0
Job Description                0
Minimum Qual Requirements     20
Preferred Skills             393
dtype: int64

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2946 entries, 0 to 2945
Data columns (total 69 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   id                         2946 non-null   int64  
 1   Target A                   2946 non-null   float64
 2   Target B                   2946 non-null   int64  
 3   LOAN_AMT                   2946 non-null   int64  
 4   Business Title             2946 non-null   object 
 5   Civil Service Title        2946 non-null   object 
 6   Division/Work Unit         2946 non-null   object 
 7   Job Description            2946 non-null   object 
 8   Minimum Qual Requirements  2926 non-null   object 
 9   Preferred Skills           2553 non-null   object 
 10  Additional Information     1854 non-null   object 
 11  To Apply                   2945 non-null   object 
 12  Hours/Shift                884 non-null    object 
 13  Residency Requirement      2942 non-null   objec

In [5]:

def missing_values_table(df):
   
    # Total missing values
    mis_val = df.isnull().sum()
    
    # Percentage of missing values
    mis_val_percent = 100 * df.isnull().sum() / len(df)
    
    # Make a table with the results
    mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
    
    # Rename the columns
    mis_val_table_columns = mis_val_table.rename(
    columns = {0 : 'Missing Values', 1 : '% of Total Values'})
    
    # Sort the table by percentage of missing descending
    # .iloc[:, 1]!= 0: filter on missing missing values not equal to zero
    mis_val_table_columns = mis_val_table_columns[
        mis_val_table_columns.iloc[:,1] != 0].sort_values(
    '% of Total Values', ascending=False).round(2)  # round(2), keep 2 digits
    
    # Print some summary information
    print("Dataset has {} columns.".format(df.shape[1]) + '\n' + 
    "There are {} columns that have missing values.".format(mis_val_table_columns.shape[0]))
    return mis_val_table_columns

In [6]:

missing_values_table(df)

Dataset has 69 columns.
There are 48 columns that have missing values.


Missing Values  % of Total Values
TimeAfterBootcamp                    2725              92.50
next_pymnt_d                         2610              88.59
HackathonReasons                     2097              71.18
Hours/Shift                          2062              69.99
FrameworkWorkedWith                  1313              44.57
FrameworkDesireNextYear              1215              41.24
SelfTaughtTypes                      1134              38.49
DatabaseDesireNextYear               1115              37.85
Additional Information               1092              37.07
PlatformDesireNextYear               1037              35.20
Methodology                           955              32.42
AIDangerous                           946              32.11
CommunicationTools                    923              31.33
Age                                   919              31.19
TimeFullyProductive                   894              30.35
AIResponsible                         876              29.74
PlatformWorkedWith                    876              29.74
AIInteresting                         870              29.53
UpdateCV                              857              29.09
EducationTypes                        818              27.77
DatabaseWorkedWith                    816              27.70
AgreeDisagree2                        813              27.60
AgreeDisagree1                        813              27.60
AgreeDisagree3                        807              27.39
AIFuture                              750              25.46
LanguageDesireNextYear                667              22.64
CheckInCode                           627              21.28
VersionControl                        614              20.84
IDE                                   586              19.89
OperatingSystem                       569              19.31
LanguageWorkedWith                    499              16.94
HopeFiveYears                         469              15.92
LastNewJob                            447              15.17
JobSearchStatus                       435              14.77
YearsCodingProf                       408              13.85
Preferred Skills                      393              13.34
UndergradMajor                        322              10.93
emp_length                            133               4.51
title                                  64               2.17
FormalEducation                        52               1.77
DevType                                45               1.53
Minimum Qual Requirements              20               0.68
Process Date                            4               0.14
Posting Updated                         4               0.14
Posting Date                            4               0.14
Residency Requirement                   4               0.14
YearsCoding                             3               0.10
To Apply                                1               0.03

In [7]:
#Dropping the columns with missing values% >60 & unnecessary columns
X1 = df.drop(['Hours/Shift','HackathonReasons','TimeAfterBootcamp','next_pymnt_d','id','Target A','Target B','Posting Date','Posting Updated','Process Date'],axis = 1)
X1.head(3)


LOAN_AMT  ... last_credit_pull_d
0      3600  ...             19-Mar
1     24700  ...             19-Mar
2     20000  ...             19-Mar

[3 rows x 59 columns]

In [8]:
X1.columns

Index(['LOAN_AMT', 'Business Title', 'Civil Service Title',
       'Division/Work Unit', 'Job Description', 'Minimum Qual Requirements',
       'Preferred Skills', 'Additional Information', 'To Apply',
       'Residency Requirement', 'FormalEducation', 'UndergradMajor',
       'CompanySize', 'DevType', 'YearsCoding', 'YearsCodingProf',
       'HopeFiveYears', 'JobSearchStatus', 'LastNewJob', 'UpdateCV',
       'CommunicationTools', 'TimeFullyProductive', 'EducationTypes',
       'SelfTaughtTypes', 'AgreeDisagree1', 'AgreeDisagree2', 'AgreeDisagree3',
       'LanguageWorkedWith', 'LanguageDesireNextYear', 'DatabaseWorkedWith',
       'DatabaseDesireNextYear', 'PlatformWorkedWith',
       'PlatformDesireNextYear', 'FrameworkWorkedWith',
       'FrameworkDesireNextYear', 'IDE', 'OperatingSystem', 'Methodology',
       'VersionControl', 'CheckInCode', 'AIDangerous', 'AIInteresting',
       'AIResponsible', 'AIFuture', 'Age', 'term', 'emp_length', 'loan_status',
       'pymnt_plan', 'url', 

In [9]:
#Imputing the missing values
from sklearn.impute import SimpleImputer
imp = SimpleImputer(missing_values=np.nan, strategy='most_frequent')
imp = imp.fit(X1)
X2 = imp.transform(X1)


In [10]:
x3= pd.DataFrame(X2)
x3.head(3)

0                                                  1   ...       57      58
0   3600                                    Account Manager  ...   122.67  19-Mar
1  24700           EXECUTIVE DIRECTOR, BUSINESS DEVELOPMENT  ...   926.35  19-Mar
2  20000  Maintenance Worker - Technical Services-Heatin...  ...  15813.3  19-Mar

[3 rows x 59 columns]

In [11]:
x3.columns = ['LOAN_AMT', 'Business Title', 'Civil Service Title',
       'Division/Work Unit', 'Job Description', 'Minimum Qual Requirements',
       'Preferred Skills', 'Additional Information', 'To Apply',
       'Residency Requirement', 'FormalEducation', 'UndergradMajor',
       'CompanySize', 'DevType', 'YearsCoding', 'YearsCodingProf',
       'HopeFiveYears', 'JobSearchStatus', 'LastNewJob', 'UpdateCV',
       'CommunicationTools', 'TimeFullyProductive', 'EducationTypes',
       'SelfTaughtTypes', 'AgreeDisagree1', 'AgreeDisagree2', 'AgreeDisagree3',
       'LanguageWorkedWith', 'LanguageDesireNextYear', 'DatabaseWorkedWith',
       'DatabaseDesireNextYear', 'PlatformWorkedWith',
       'PlatformDesireNextYear', 'FrameworkWorkedWith',
       'FrameworkDesireNextYear', 'IDE', 'OperatingSystem', 'Methodology',
       'VersionControl', 'CheckInCode', 'AIDangerous', 'AIInteresting',
       'AIResponsible', 'AIFuture', 'Age', 'term', 'emp_length', 'loan_status',
       'pymnt_plan', 'url', 'purpose', 'title', 'zip_code', 'addr_state',
       'dti', 'earliest_cr_line', 'last_pymnt_d', 'last_pymnt_amnt',
       'last_credit_pull_d']

In [12]:
x3.isnull().sum()

LOAN_AMT                     0
Business Title               0
Civil Service Title          0
Division/Work Unit           0
Job Description              0
Minimum Qual Requirements    0
Preferred Skills             0
Additional Information       0
To Apply                     0
Residency Requirement        0
FormalEducation              0
UndergradMajor               0
CompanySize                  0
DevType                      0
YearsCoding                  0
YearsCodingProf              0
HopeFiveYears                0
JobSearchStatus              0
LastNewJob                   0
UpdateCV                     0
CommunicationTools           0
TimeFullyProductive          0
EducationTypes               0
SelfTaughtTypes              0
AgreeDisagree1               0
AgreeDisagree2               0
AgreeDisagree3               0
LanguageWorkedWith           0
LanguageDesireNextYear       0
DatabaseWorkedWith           0
DatabaseDesireNextYear       0
PlatformWorkedWith           0
Platform

In [13]:
x3.head(5)

LOAN_AMT  ... last_credit_pull_d
0     3600  ...             19-Mar
1    24700  ...             19-Mar
2    20000  ...             19-Mar
3    35000  ...             19-Mar
4    10400  ...             18-Mar

[5 rows x 59 columns]

In [14]:
print('No. of missing values in Target A is {0}'.format(df['Target A'].isnull().sum()))
print('No. of missing values in Target B is {0}'.format(df['Target B'].isnull().sum()))

No. of missing values in Target A is 0
No. of missing values in Target B is 0


In [15]:
import seaborn as sns


In [ ]:
for i in x3.iloc[:,1:].columns:
  sns.boxplot(y=df['Target B'],x=df[i])
  plt.show()





In [16]:
x4 = pd.get_dummies(x3,columns=['Business Title', 'Civil Service Title',
       'Division/Work Unit', 'Job Description', 'Minimum Qual Requirements',
       'Preferred Skills', 'Additional Information', 'To Apply',
       'Residency Requirement', 'FormalEducation', 'UndergradMajor',
       'CompanySize', 'DevType', 'YearsCoding', 'YearsCodingProf',
       'HopeFiveYears', 'JobSearchStatus', 'LastNewJob', 'UpdateCV',
       'CommunicationTools', 'TimeFullyProductive', 'EducationTypes',
       'SelfTaughtTypes', 'AgreeDisagree1', 'AgreeDisagree2', 'AgreeDisagree3',
       'LanguageWorkedWith', 'LanguageDesireNextYear', 'DatabaseWorkedWith',
       'DatabaseDesireNextYear', 'PlatformWorkedWith',
       'PlatformDesireNextYear', 'FrameworkWorkedWith',
       'FrameworkDesireNextYear', 'IDE', 'OperatingSystem', 'Methodology',
       'VersionControl', 'CheckInCode', 'AIDangerous', 'AIInteresting',
       'AIResponsible', 'AIFuture', 'Age', 'term', 'emp_length', 'loan_status',
       'pymnt_plan', 'url', 'purpose', 'title', 'zip_code', 'addr_state',
       'dti', 'earliest_cr_line', 'last_pymnt_d', 'last_pymnt_amnt',
       'last_credit_pull_d'])


In [17]:
print('Shape before encoding is {0} & after encoding is {1}'.format(x3.shape,x4.shape))

Shape before encoding is (2946, 59) & after encoding is (2946, 25627)


In [18]:
X = x4.values
y = df['Target B'].values
y1 = df['Target A'].values

#For Target B

In [19]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.20,random_state=0)


In [36]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [37]:
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression

In [27]:
models = []
models.append(('Logistic Regression',LogisticRegression(random_state=0)))
models.append(('Naive Bayes',GaussianNB()))
models.append(('RandomForest', RandomForestClassifier(n_estimators=10,random_state=0)))
models.append(('Decision Tree', DecisionTreeClassifier(random_state=0)))
models.append(('KNN', KNeighborsClassifier(n_neighbors = 30)))

In [29]:
for name, model in models:
    print(name)
    model.fit(X_train, y_train)
    #Checking the cross_val_score
    score=cross_val_score(model,X,y,cv=10)
    print('Cross Val score is {0}'.format(score.mean()))
    
    
    # Make predictions.
    predictions = model.predict(X_test)
    # Compute the error.
    from sklearn.metrics import confusion_matrix
    print(confusion_matrix(predictions, y_test))
    print('\n')
    from sklearn.metrics import accuracy_score
    print('Accuracy Score: {0}'.format(accuracy_score(predictions,y_test)))
    print('\n')

Logistic Regression


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


Cross Val score is 0.7878508013374841
[[417 118]
 [ 38  17]]


Accuracy Score: 0.735593220338983


Naive Bayes
Cross Val score is 0.5780998501095354
[[226  61]
 [229  74]]


Accuracy Score: 0.5084745762711864


RandomForest
Cross Val score is 0.7885276144356047
[[454 134]
 [  1   1]]


Accuracy Score: 0.7711864406779662


Decision Tree
Cross Val score is 0.7189380837080596
[[402 117]
 [ 53  18]]


Accuracy Score: 0.711864406779661


KNN
Cross Val score is 0.7898846996425689
[[455 135]
 [  0   0]]


Accuracy Score: 0.7711864406779662




In [32]:
for name,classifier in models:
  y_pred = classifier.predict(X_test)
  f1_sc = f1_score(y_test,y_pred)
  print(name,'-',f1_sc)
  

Logistic Regression - 0.17894736842105263
Naive Bayes - 0.3378995433789954
RandomForest - 0.014598540145985401
Decision Tree - 0.17475728155339806
KNN - 0.0


#Best F_score till now
####Decision Tree:0.17475728155339806
####Naive Bayes:0.3378995433789954
####logistic Regression:0.17894736842105263

#For Target A

In [ ]:
y2 = df.astype({'Target A':int})
y2.head()
y3 = y2.iloc[:,1]
y3.shape
y3.dtype



dtype('int64')

In [1]:
X_train,X_test,y_train1,y_test1 = train_test_split(X,y3,test_size=0.20,random_state=0)


NameError: ignored

In [ ]:
y_train1.shape

(2356,)

In [ ]:
from sklearn.tree import DecisionTreeRegressor
classifier1 = DecisionTreeRegressor(random_state=0)
classifier1.fit(X_train, y_train1)

DecisionTreeRegressor(ccp_alpha=0.0, criterion='mse', max_depth=None,
                      max_features=None, max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, presort='deprecated',
                      random_state=0, splitter='best')

In [ ]:
from sklearn.metrics import confusion_matrix,accuracy_score,f1_score,mean_squared_error     

In [ ]:
y2_pred = classifier1.predict(X_test)

print("The mean squared error is {0}".format(mean_squared_error(y_test1,y2_pred,squared=False)))



The mean squared error is 22504.597705388023


In [ ]:
from sklearn.svm import SVR
regressor = SVR(kernel = 'rbf')
regressor.fit(X_train, y_train1)

SVR(C=1.0, cache_size=200, coef0=0.0, degree=3, epsilon=0.1, gamma='scale',
    kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False)

In [ ]:
y3_pred = regressor.predict(X_test)

In [ ]:
print("The mean squared error is {0}".format(mean_squared_error(y_test1,y3_pred,squared=False)))

The mean squared error is 27153.4676065291
